cmd/backup/script.go

```diff
@@ -369,38 +436,111 @@ func (s *script) stopContainers() (func() error, error) {
 		}
 	}
 
-	var stopError error
-	if len(stopErrors) != 0 {
-		stopError = fmt.Errorf(
-			"stopContainers: %d error(s) stopping containers: %w",
-			len(stopErrors),
-			errors.Join(stopErrors...),
-		)
+	var scaledDownServices []swarm.Service
+	var scaleDownErrors []error
+	if isDockerSwarm {
+		for _, svc := range servicesToScaleDown {
```

Comment:
I'm wondering if both scaling down and scaling up multiple services should happen concurrently as it can take quite a while for a single service to converge, making downtimes during backup a bit longer than needed.

Then again we already have problems like described in https://github.com/offen/docker-volume-backup/issues/160

Any opinions on this @pixxon ?

Reply:
Since after scale down there is a check to ensure that every container is fully removed, it should be possible to initiate the scale down on every service in a detached mode. ( I am not sure if the go api has this feature, but the cli does. )

Regarding the scale up, it could also be done in a detached mode, since backup can move on from there. Only thing that could be helpful is to check if every service is up to expected number before a timeout limit. And if not, issue a warning so the user is aware.

Reply:
As for __scaling up__: I'm not sure if detaching would actually make this faster. If you look at the code in Docker CLI: https://github.com/docker/cli/blob/abf8cff233f6d3bfb93d0c15f5a55f1987418ae3/cli/command/service/scale.go#L80-L88 it seems detaching skips what we are doing here when calling `progress.ServiceProgress`. However, the loop that polls for no containers to be present anymore will apparently always takes longer than `progress.ServiceProgress`. So all that "detaching" (i.e. skipping `progress.ServiceProgress`) would give is us fewer meaningful errors (as returned by `progress.ServiceProgress` and the possibility of getting stuck in an endless loop).

As for __scaling down__: Detaching would indeed make things considerably faster, then again this would not allow notification features to kick in in the case of failures to scale back up. Maybe a simple solution would be to run the entire "restart" function in a goroutine so that encryption, upload et al can happen in parallel.

Reply:
I think the detach would still help you with the scaling up, since the order of the scales and waits would be different. ( In attached mode: `scale wait scale wait scale wait` in detached mode: `scale scale scale wait wait wait`.

To me really the only difference is that if one of the services fail to scale down, we could terminate earlier. But the users should not rely on this, since it is non-deterministic.

Reply:
> scale scale scale wait wait wait

This is exactly what we'd get if every operation on a service would be running in its own goroutine (as implemented in cfa5f073d2b31b3b64e7d81108f4ce905885119a)

Reply:
Then I am not exactly sure why you mean it would not make it faster. The reordering should allow Docker to work in the background while we move on to the next service.

Regarding the endless loop, I think one reasonable solution would be to set a timeout limit, after which the backup will just assume there is an error. Sadly I don't think we can really do much about this, since Docker itself will just have a container stuck in Terminating or some other state for a longer period of time.

Reply:
> Then I am not exactly sure why you mean it would not make it faster. 

I think this is mostly a misunderstanding. It's running in parallel now in any case, not trying to bail in case a service fails, but trying to finish all of them. I think that's a bit easier to understand.

> Regarding the endless loop, I think one reasonable solution would be to set a timeout limit, after which the backup will just assume there is an error. 

Added it like this with a 5 minute timeout.